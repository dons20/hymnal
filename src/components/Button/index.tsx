import { Button as CoreButton, ButtonProps } from "@chakra-ui/react";
import { Link, LinkProps } from "react-router-dom";

interface ButtonLink extends ButtonProps {
	as: typeof Link;
}

type ButtonP = ButtonProps | (ButtonLink & LinkProps);

const Button = ({ children, size = "lg", colorScheme = "blue", ...rest }: ButtonP) => (
	<CoreButton size={size} colorScheme={colorScheme} {...rest}>
		{children}
	</CoreButton>
);

export default Button;
