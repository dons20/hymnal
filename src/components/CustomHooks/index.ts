import useSongLoader from "./useSongLoader";

export { useSongLoader };

export default null;
