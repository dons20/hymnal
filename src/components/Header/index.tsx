import React, { lazy, useState } from "react";
import { useColorMode, useMediaQuery, useDisclosure, useColorModeValue, Box } from "@chakra-ui/react";
import { FaSearch, FaSun, FaMoon, FaHome, FaBars } from "react-icons/fa";
import { useDebouncedCallback } from "use-debounce";
import withSuspense from "helpers/withSuspense";
import { useMainContext } from "utils/context";
import { useNavigate } from "react-router-dom";
import Fuse, { FuseResult } from "fuse.js";
import "./Header.scss";

import type {
	Icon as IconType,
	Grid as GridType,
	Text as TextType,
	Fade as FadeType,
	Input as InputType,
	Modal as ModalType,
	Portal as PortalType,
	VStack as VStackType,
	Heading as HeadingType,
	ModalBody as ModalBodyType,
	IconButton as IconButtonType,
	InputGroup as InputGroupType,
	ModalHeader as ModalHeaderType,
	CloseButton as CloseButtonType,
	ModalOverlay as ModalOverlayType,
	ModalContent as ModalContentType,
	ModalCloseButton as ModalCloseButtonType,
	InputRightElement as InputRightElementType,
} from "@chakra-ui/react/dist";

/* Lazy Base Imports */
const CloseButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.CloseButton })));
const FadeImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Fade })));
const GridImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Grid })));
const HeadingImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Heading })));
const InputImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Input })));
const IconImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Icon })));
const IconButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.IconButton })));
const InputGroupImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.InputGroup })));
const InputRightElementImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.InputRightElement })));
const ModalImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Modal })));
const ModalOverlayImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalOverlay })));
const ModalContentImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalContent })));
const ModalHeaderImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalHeader })));
const ModalCloseButtonImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalCloseButton })));
const ModalBodyImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.ModalBody })));
const PortalImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Portal })));
const TextImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.Text })));
const VStackImport = lazy(() => import("@chakra-ui/react").then(m => ({ default: m.VStack })));
const ButtonImport = lazy(() => import("components/Button"));

/* With Suspense Wrapper */
const Fade = withSuspense<typeof FadeType>(FadeImport);
const Grid = withSuspense<typeof GridType, null>(GridImport, null);
const Heading = withSuspense<typeof HeadingType, null>(HeadingImport, null);
const Input = withSuspense<typeof InputType, null>(InputImport, null);
const Icon = withSuspense<typeof IconType>(IconImport);
const CloseButton = withSuspense<typeof CloseButtonType>(CloseButtonImport);
const IconButton = withSuspense<typeof IconButtonType>(IconButtonImport);
const InputGroup = withSuspense<typeof InputGroupType, null>(InputGroupImport, null);
const InputRightElement = withSuspense<typeof InputRightElementType, null>(InputRightElementImport, null);
const Modal = withSuspense<typeof ModalType, null>(ModalImport, null);
const ModalOverlay = withSuspense<typeof ModalOverlayType, null>(ModalOverlayImport, null);
const ModalContent = withSuspense<typeof ModalContentType, null>(ModalContentImport, null);
const ModalHeader = withSuspense<typeof ModalHeaderType, null>(ModalHeaderImport, null);
const ModalCloseButton = withSuspense<typeof ModalCloseButtonType, null>(ModalCloseButtonImport, null);
const ModalBody = withSuspense<typeof ModalBodyType, null>(ModalBodyImport, null);
const Portal = withSuspense<typeof PortalType, null>(PortalImport, null);
const Text = withSuspense<typeof TextType>(TextImport);
const VStack = withSuspense<typeof VStackType, null>(VStackImport, null);
const Button = withSuspense<typeof ButtonImport, null>(ButtonImport, null);

function Header() {
	const navigate = useNavigate();
	const { songs } = useMainContext();
	const { colorMode, toggleColorMode } = useColorMode();
	const { isOpen, onOpen, onClose } = useDisclosure();
	const { isOpen: isModalOpen, onOpen: onModalOpen, onClose: onModalClose } = useDisclosure();
	const fuse = new Fuse(songs!, { keys: ["number", "title"], minMatchCharLength: 2, threshold: 0.4 });
	const [query, setQuery] = useState("");
	const [mobileQuery, setMobileQuery] = useState("");
	const [queryResults, setQueryResults] = useState<FuseResult<Song>[]>([]);
	const [showMobileMenu] = useMediaQuery("(max-width: 550px)");
	const resultsBG = useColorModeValue("white", "gray.800");
	const headerBG = useColorModeValue("gray.100", "gray.800");
	const searchBG = useColorModeValue("gray.50", "gray.700");

	/** Will navigate one level up in the application */
	const GoHome = () => navigate("/");

	const searchSongs = (_?: React.ChangeEvent<any>, mobile?: boolean) => {
		const shouldSearchMobile = mobile && mobileQuery.length > 0;
		const shouldSearchDesktop = !mobile && query.length > 0;
		if (shouldSearchMobile || shouldSearchDesktop) {
			onClose();
			onModalClose();
			setQuery("");
			setMobileQuery("");
			setQueryResults([]);
			navigate(`/search?query=${shouldSearchMobile ? mobileQuery : query}`);
		}
	};

	const submitQuery = (e: React.FormEvent<HTMLDivElement>) => {
		e.preventDefault();
		if (query.length > 0) searchSongs();
	};
	const submitMobileQuery = (e: React.FormEvent<HTMLDivElement>) => {
		e.preventDefault();
		if (mobileQuery.length > 0) searchSongs(e, true);
	};

	const mobileSearchQueryChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const searchValue = e.target.value;
		setMobileQuery(searchValue);
	};

	const handleSearch = useDebouncedCallback((value: string) => {
		if (value.length === 0) {
			onClose();
			return;
		}
		const result = fuse.search(value, { limit: 6 });
		setQueryResults(result.slice(0, 6));
		if (result.length > 0) onOpen();
	}, 500);

	const searchQueryChange = (value: string) => {
		setQuery(value);
		handleSearch(value);
	};

	const gotoSong = (index: number) => {
		onClose();
		setQuery("");
		navigate(`/songs/${index}`);
	};

	return (
		<Box className="page-header" p={4} bg={headerBG}>
			<Grid templateColumns="auto 1fr" alignItems="center" gap={10} justifyContent="space-between" px={5}>
				<Heading size="md" onClick={GoHome} cursor="pointer" display="flex" alignItems="center" width="auto">
					Hymns for All Times
					<Icon as={FaHome} size={20} ml={3} />
				</Heading>
				{showMobileMenu ? (
					<IconButton
						size="xs"
						icon={<FaBars />}
						aria-label="Open mobile menu"
						justifySelf="flex-end"
						onClick={onModalOpen}
						data-testid="mobileMenuTrigger"
					/>
				) : (
					<Grid templateColumns="minmax(auto, 300px) auto" gap={2} justifyContent="flex-end">
						<InputGroup size="md" as="form" onSubmit={submitQuery} role="search">
							<Input
								value={query}
								onChange={(e: React.ChangeEvent<HTMLInputElement>) => searchQueryChange(e.target.value)}
								type="search"
								placeholder="Search songs..."
								pr="4.5rem"
								bg={searchBG}
								data-testid="desktopSearch"
							/>
							<InputRightElement>
								<IconButton
									size="sm"
									h="1.75rem"
									icon={<FaSearch />}
									aria-label="Search Song Database"
									onClick={searchSongs}
								/>
							</InputRightElement>
						</InputGroup>

						<IconButton
							size="md"
							w="1.5rem"
							onClick={toggleColorMode}
							icon={colorMode === "light" ? <FaMoon /> : <FaSun />}
							aria-label="Toggle Color Mode"
						/>
					</Grid>
				)}
			</Grid>
			<Fade in={isOpen}>
				<Portal>
					<Box
						bg={resultsBG}
						pos="absolute"
						zIndex={105}
						top="60px"
						right="65"
						w={350}
						px={5}
						pb={5}
						hidden={!isOpen || showMobileMenu || undefined}
						borderRadius="md"
						shadow="md"
						data-testid="searchResultsBox"
					>
						<CloseButton size="md" onClick={onClose} pos="absolute" top="1" right="5" />
						<Grid gap={5} mt={10} w="100%" data-testid="searchItemsWrapper">
							{queryResults.map(result => (
								<Button
									bg="blue.500"
									onClick={() => gotoSong(result.item.number)}
									overflow="hidden"
									key={result.item.number}
									role="button"
								>
									<Grid templateColumns="auto 1fr" gap="3" w="100%" justifyItems="left">
										<Text size="sm" color="white">
											{result.item.number}
										</Text>
										<Text size="sm" color="white">
											{result.item.title}
										</Text>
									</Grid>
								</Button>
							))}
						</Grid>
					</Box>
				</Portal>
			</Fade>

			<Modal isOpen={isModalOpen} onClose={onModalClose}>
				<ModalOverlay />
				<ModalContent pb={10}>
					<ModalHeader>Menu</ModalHeader>
					<ModalCloseButton />
					<ModalBody>
						<VStack spacing="4">
							<InputGroup size="md" as="form" onSubmit={submitMobileQuery}>
								<Input
									type="search"
									value={mobileQuery}
									onChange={mobileSearchQueryChange}
									placeholder="Search songs..."
									pr="4.5rem"
									bg={searchBG}
									data-testid="mobileSearch"
								/>
								<InputRightElement>
									<IconButton
										size="sm"
										h="1.75rem"
										icon={<FaSearch />}
										aria-label="Search Song Database"
										onClick={(e: React.MouseEvent<HTMLButtonElement>) => searchSongs(e, true)}
									/>
								</InputRightElement>
							</InputGroup>

							<Button
								leftIcon={colorMode === "light" ? <FaMoon /> : <FaSun />}
								onClick={toggleColorMode}
								w="100%"
							>
								Toggle {colorMode === "light" ? "Dark" : "Light"} Mode
							</Button>
						</VStack>
					</ModalBody>
				</ModalContent>
			</Modal>
		</Box>
	);
}

export default Header;
