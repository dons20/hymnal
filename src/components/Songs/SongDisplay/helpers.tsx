import React from 'react';
import { Box, Text } from '@mantine/core';

/**
 * Helper function to render chorus with vocal parts and indicators
 */
export const renderChorusWithParts = (song: Song, isDark: boolean) => {
  if (!song.chorusParts || song.chorusParts.length === 0) {
    // Fallback to regular chorus rendering
    return <Text className="song-text">{song.chorus}</Text>;
  }

  return (
    <Box className="chorus-parts">
      {song.chorusParts.map((part, index) => (
        <Box
          key={index}
          className="chorus-part"
          style={{ position: 'relative', marginBottom: '0.5rem' }}
        >
          {part.part && (
            <Text
              component="span"
              className="part-indicator-regular"
              style={
                {
                  '--part-indicator-color':
                    part.part === 'soprano'
                      ? isDark
                        ? '#74c0fc'
                        : '#1c7ed6' // Blue for soprano
                      : isDark
                        ? '#ffec99'
                        : '#fd7e14', // Orange for baritone
                  '--part-indicator-bg':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.15)'
                        : 'rgba(28, 126, 214, 0.08)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.15)'
                        : 'rgba(253, 126, 20, 0.08)',
                  '--part-indicator-border':
                    part.part === 'soprano'
                      ? isDark
                        ? 'rgba(116, 192, 252, 0.3)'
                        : 'rgba(28, 126, 214, 0.25)'
                      : isDark
                        ? 'rgba(255, 236, 153, 0.3)'
                        : 'rgba(253, 126, 20, 0.25)',
                } as React.CSSProperties
              }
            >
              {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
            </Text>
          )}
          <Text
            className="song-text"
            style={
              {
                paddingLeft: part.part ? '0' : '0',
                fontWeight: part.part === 'soprano' ? 500 : 400,
                '--part-text-color':
                  part.part === 'soprano'
                    ? isDark
                      ? '#e9ecef'
                      : '#212529' // Standard text color for soprano
                    : isDark
                      ? '#ced4da'
                      : '#495057', // Slightly muted for baritone
                color: 'var(--part-text-color)',
              } as React.CSSProperties
            }
          >
            {part.text}
          </Text>
        </Box>
      ))}
    </Box>
  );
};

/**
 * Helper function to render presentation slide content (verses or chorus with vocal parts)
 */
export const renderPresentationContent = (content: string | VocalPart[], isDark: boolean) => {
  if (typeof content === 'string') {
    return <Text className="presentation-text">{content}</Text>;
  }

  // Handle VocalPart[] for presentation mode with larger styling
  return (
    <Box className="presentation-content-container">
      {content.map((part, index) => {
        // Check if this is the end of a B part followed by an A part (new pair starting)
        const isEndOfPair =
          part.part === 'baritone' &&
          index < content.length - 1 &&
          content[index + 1]?.part === 'soprano';

        return (
          <React.Fragment key={index}>
            <Box
              className="presentation-part-container"
              style={{
                marginBottom: isEndOfPair ? '2rem' : '1.5rem',
              }}
            >
              {part.part && (
                <Text
                  component="span"
                  className="part-indicator-presentation"
                  style={
                    {
                      '--part-indicator-color':
                        part.part === 'soprano'
                          ? isDark
                            ? '#74c0fc'
                            : '#1c7ed6' // Blue for soprano
                          : isDark
                            ? '#ffec99'
                            : '#fd7e14', // Orange for baritone
                      '--part-indicator-bg':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.15)'
                            : 'rgba(28, 126, 214, 0.08)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.15)'
                            : 'rgba(253, 126, 20, 0.08)',
                      '--part-indicator-border':
                        part.part === 'soprano'
                          ? isDark
                            ? 'rgba(116, 192, 252, 0.3)'
                            : 'rgba(28, 126, 214, 0.25)'
                          : isDark
                            ? 'rgba(255, 236, 153, 0.3)'
                            : 'rgba(253, 126, 20, 0.25)',
                    } as React.CSSProperties
                  }
                >
                  {part.part === 'soprano' ? 'A' : part.part === 'baritone' ? 'B' : '♪'}
                </Text>
              )}
              <Text
                className="presentation-part-text"
                style={
                  {
                    fontWeight: part.part === 'soprano' ? 600 : 400,
                    '--part-text-color':
                      part.part === 'soprano'
                        ? isDark
                          ? '#e9ecef'
                          : '#212529' // Standard text color for soprano
                        : isDark
                          ? '#ced4da'
                          : '#495057', // Slightly muted for baritone
                  } as React.CSSProperties
                }
              >
                {part.text}
              </Text>
            </Box>
            {isEndOfPair && (
              <Box
                className="pair-divider"
                style={
                  {
                    '--divider-color': isDark
                      ? 'rgba(255, 255, 255, 0.2)'
                      : 'rgba(0, 0, 0, 0.15)',
                  } as React.CSSProperties
                }
              />
            )}
          </React.Fragment>
        );
      })}
    </Box>
  );
};
