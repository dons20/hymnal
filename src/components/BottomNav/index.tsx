import React, { useState, useEffect, useRef } from "react";
import { FaItunesNote, FaHome, FaStar } from "react-icons/fa";
import { useLocation, useNavigate } from "react-router-dom";
import { useColorModeValue } from "@chakra-ui/color-mode";
import { useMediaQuery } from "@chakra-ui/media-query";
import { useDebouncedCallback } from "use-debounce";
import { isMobile } from "react-device-detect";
import { Box, Text } from "@chakra-ui/layout";
import Button from "components/Button";
import cx from "classnames";
import "./BottomNav.scss";

function MobileNavBar() {
	const { pathname } = useLocation();
	const navigate = useNavigate();
	const [scrollingDown, setScrollingDown] = useState(true);
	const prevPath = useRef<string | null>(null);
	const scrollPos = useRef(document.body.getBoundingClientRect().top);
	const [iconsOnly] = useMediaQuery("(max-width: 475px)");
	const footerBg = useColorModeValue("blue.600", "blue.600");
	const footerColors = useColorModeValue("gray.100", "gray.100");

	const tabValues = [
		{
			title: "Home",
			url: "/home",
			icon: <FaHome />,
		},
		{
			title: "Songs",
			url: "/songs/index",
			icon: <FaItunesNote />,
		},
		{
			title: "Favourites",
			url: "/songs/favourites",
			icon: <FaStar />,
		},
	];

	const scrollEventHandler = useDebouncedCallback(() => {
		// If we're within 100px of the bottom, show the bottom nav regardless of scroll direction
		const windowHeight = Math.round(window.innerHeight);
		const documentHeight = Math.round(document.body.offsetHeight);
		const scrollY = Math.round(window.scrollY);
		
		if (scrollY >= (documentHeight - windowHeight) - 100) {
			setScrollingDown(false);
		// detects new state and compares it with the new one
		} else if (document.body.getBoundingClientRect().top > scrollPos.current) {
			setScrollingDown(false);
		} else setScrollingDown(true);

		// saves the new position for iteration.
		scrollPos.current = Math.round(document.body.getBoundingClientRect().top);
	}, 100, { leading: true, trailing: true });

	useEffect(() => {
		scrollPos.current = Math.round(document.body.getBoundingClientRect().top);
		prevPath.current = pathname;
		setScrollingDown(true);

		// adding scroll event
		window.addEventListener("scroll", scrollEventHandler);

		return () => {
			window.removeEventListener("scroll", scrollEventHandler);
		};
	}, [setScrollingDown, pathname, scrollEventHandler]);

	function handleTabBarPress(e: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
		const url = e.currentTarget.getAttribute("data-url") as string;
		navigate(url);
	}

	return (
		<Box
			className={cx("bottom-nav", { "--disabled": !isMobile, "--hidden": scrollingDown })}
			color={footerColors}
			bg={footerBg}
			data-testid="bottomNavWrapper"
		>
			{tabValues.map(tab => (
				<Button
					data-url={tab.url}
					leftIcon={!iconsOnly ? tab.icon : undefined}
					aria-label={tab.title}
					onClick={handleTabBarPress}
					key={tab.title}
					color={footerColors}
					bg="transparent"
					px={10}
					data-testid={tab.title}
				>
					{iconsOnly && tab.icon}
					{!iconsOnly && <Text>{tab.title}</Text>}
				</Button>
			))}
		</Box>
	);
}

export default MobileNavBar;
