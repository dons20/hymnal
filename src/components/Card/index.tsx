import { useColorModeValue } from "@chakra-ui/color-mode";
import { Box, Image } from "@chakra-ui/react";
import { Link } from "react-router-dom";
import Button from "components/Button";
import Loader from "components/Loader";

interface CardBase {
	/** A title for the card */
	title: string;
	/** A subtitle to display next to the title */
	subtitle: string;
	/** The label for the primary link */
	primaryLabel: string;
	/** A route to navigate to when clicking the primary button */
	primaryLink: string;
}

interface CardWithImage extends CardBase {
	/** The source for the card image */
	imageSrc: string;
	/** The alt text for the image (accessibility) */
	imageAlt: string;
}

interface CardWithoutImage extends CardBase {
	imageSrc?: never;
	imageAlt?: never;
}

type CardP = CardWithImage | CardWithoutImage;

const Card = ({ title, subtitle, primaryLink, primaryLabel, imageSrc, imageAlt }: CardP) => {
	const boxBG = useColorModeValue("white", "gray.600");
	const subtitleColor = useColorModeValue("gray.600", "gray.200");

	return (
		<Box className="card" bg={boxBG} p="4" shadow="md" borderRadius="sm">
			<Box display="flex" justifyContent="space-between" alignItems="center" flexWrap="wrap">
				<Box>{title}</Box>
				<Box color={subtitleColor} fontSize="sm" as="span">
					{subtitle}
				</Box>
			</Box>
			{imageSrc && (
				<Image
					loading="lazy"
					boxSize="280px"
					objectFit="contain"
					fallback={<Loader />}
					alt={imageAlt}
					src={imageSrc}
					htmlHeight="300px"
					htmlWidth="300px"
					h="100%"
					w="100%"
					marginBottom="6"
					minH={280}
				/>
			)}
			<Button size="md" colorScheme="blue" as={Link} to={primaryLink}>
				{primaryLabel}
			</Button>
		</Box>
	);
};

export default Card;
