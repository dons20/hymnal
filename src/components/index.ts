import Card from "./Card";
import Loader from "./Loader";
import Header from "./Header";
import Button from "./Button";
import BottomNav from "./BottomNav";
import PictureHeader from "./PictureHeader";
import HomepageSearch from "./HomepageSearch";

export { BottomNav, Button, Header, PictureHeader, Card, Loader, HomepageSearch };
