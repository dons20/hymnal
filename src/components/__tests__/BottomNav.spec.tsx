import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { resizeWindow } from "helpers/tests";
import * as rdd from "react-device-detect";
import { BottomNav } from "components";

// @ts-expect-error
rdd.isMobile = true;
const user = userEvent.setup();

beforeEach(() => {
    window.matchMedia = jest.fn().mockImplementation(() => ({
        matches: false,
        media: "",
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
    }));
});

describe("#BottomNav", () => {
	it("should render correctly", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<BottomNav />
			</MemoryRouter>
		);
		const Navbar = screen.getByTestId("bottomNavWrapper");
		const HomeButton = screen.getByTestId("Home");
		const SongsButton = screen.getByTestId("Songs");
		const FavouritesButton = screen.getByTestId("Favourites");

		expect(asFragment()).toMatchSnapshot();
		expect(Navbar).toBeInTheDocument();
		expect(HomeButton).toBeInTheDocument();
		expect(SongsButton).toBeInTheDocument();
		expect(FavouritesButton).toBeInTheDocument();
	});
	it("shows only icons on smaller screens", () => {
		render(
			<MemoryRouter>
				<BottomNav />
			</MemoryRouter>
		);
		resizeWindow(450, 500);
		const HomeButton = screen.getByTestId("Home");
		const SongsButton = screen.getByTestId("Songs");
		const FavouritesButton = screen.getByTestId("Favourites");

		expect(HomeButton.innerText).toBeFalsy();
		expect(SongsButton.innerText).toBeFalsy();
		expect(FavouritesButton.innerText).toBeFalsy();

		resizeWindow(768, 500);
		expect(HomeButton).toHaveTextContent("Home");
		expect(SongsButton).toHaveTextContent("Songs");
		expect(FavouritesButton).toHaveTextContent("Favourites");
	});
	it("navigates to the correct route when each button is clicked", async () => {
		render(
			<BrowserRouter>
				<BottomNav />
			</BrowserRouter>
		);
		const HomeButton = screen.getByTestId("Home");
		const SongsButton = screen.getByTestId("Songs");
		const FavouritesButton = screen.getByTestId("Favourites");

		await user.click(FavouritesButton);
		expect(window.location.href).toContain("/songs/favourites");

		await user.click(SongsButton);
		expect(window.location.href).toContain("/songs/index");

		await user.click(HomeButton);
		expect(window.location.href).toContain("/home");
	});
	it.skip("hides when scrolling up and shows when scrolling down", async () => {
		render(
			<MemoryRouter>
				<div style={{ height: 1400 }} />
				<BottomNav />
			</MemoryRouter>
		);
		fireEvent.scroll(window, { target: { scrollY: 500 } });
		const navbar = await screen.findByTestId("bottomNavWrapper");
		await waitFor(() => expect(navbar).toHaveClass("--disabled"));

		fireEvent.scroll(window, { target: { scrollY: 300 } });
		await waitFor(() => expect(navbar).toHaveClass("--disabled", "--hidden"));
	});
});
