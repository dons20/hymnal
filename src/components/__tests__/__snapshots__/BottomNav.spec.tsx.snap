// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#BottomNav should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bottom-nav --hidden css-lvpuj"
    data-testid="bottomNavWrapper"
  >
    <button
      aria-label="Home"
      class="chakra-button css-lrkawa"
      data-testid="Home"
      data-url="/home"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          />
        </svg>
      </span>
      <p
        class="chakra-text css-0"
      >
        Home
      </p>
    </button>
    <button
      aria-label="Songs"
      class="chakra-button css-lrkawa"
      data-testid="Songs"
      data-url="/songs/index"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M381.9 388.2c-6.4 27.4-27.2 42.8-55.1 48-24.5 4.5-44.9 5.6-64.5-10.2-23.9-20.1-24.2-53.4-2.7-74.4 17-16.2 40.9-19.5 76.8-25.8 6-1.1 11.2-2.5 15.6-7.4 6.4-7.2 4.4-4.1 4.4-163.2 0-11.2-5.5-14.3-17-12.3-8.2 1.4-185.7 34.6-185.7 34.6-10.2 2.2-13.4 5.2-13.4 16.7 0 234.7 1.1 223.9-2.5 239.5-4.2 18.2-15.4 31.9-30.2 39.5-16.8 9.3-47.2 13.4-63.4 10.4-43.2-8.1-58.4-58-29.1-86.6 17-16.2 40.9-19.5 76.8-25.8 6-1.1 11.2-2.5 15.6-7.4 10.1-11.5 1.8-256.6 5.2-270.2.8-5.2 3-9.6 7.1-12.9 4.2-3.5 11.8-5.5 13.4-5.5 204-38.2 228.9-43.1 232.4-43.1 11.5-.8 18.1 6 18.1 17.6.2 344.5 1.1 326-1.8 338.5z"
          />
        </svg>
      </span>
      <p
        class="chakra-text css-0"
      >
        Songs
      </p>
    </button>
    <button
      aria-label="Favourites"
      class="chakra-button css-lrkawa"
      data-testid="Favourites"
      data-url="/songs/favourites"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          />
        </svg>
      </span>
      <p
        class="chakra-text css-0"
      >
        Favourites
      </p>
    </button>
  </div>
</DocumentFragment>
`;
